src/Orleans/Async/Cancellation/CancellationSourcesExtension.cs

```diff
@@ -0,0 +1,57 @@
+﻿using System;
+using System.Threading.Tasks;
+using Orleans.Runtime;
+
+namespace Orleans.Async
+{
+    /// <summary>
+    /// Contains list of cancellation token source corresponding to the tokens
+    /// passed to the related grain activation.
+    /// </summary>
+    internal class CancellationSourcesExtension : ICancellationSourcesExtension
+    {
+        private readonly Lazy<TraceLogger> _logger = new Lazy<TraceLogger>(() =>
+            TraceLogger.GetLogger("CancellationSourcesExtension", TraceLogger.LoggerType.Application));
+        private static readonly Interner<Guid, GrainCancellationTokenSource> _cancellationTokenSources;
```

Comment:
I think I understand another confusion we are having.
This should not be `GrainCancellationTokenSource`. This should be `GrainCancellationToken`.
The GCTS (the source) can not be send around (its not serializable and should not be and you should add a condition to our serializer too throw if its attempted to be send around).
Applications creates and holds GCTS and we don't care about it.

Source points to the token.
**The Token is what can be send around remotely.**
When token arrives remotely, we store a weak ref to it here and pass to the app. 
Later on the sender calls Cancel on the GTCS. That call (just the Guid arg in the extension method) arrives into the destination and we lookup the Token by Guid. If we find it, we call Cancel on it. 

That is it. No need for us to store GCTS .